### PR TITLE
Elaborate relational errors at least once when reporting them

### LINKED ISSUE #1734: Error: Class incorrectly implements interface with no reason specified
I'm getting the following compiler error using TypeScript 1.4.1:

```
error TS2420: Class 'WorkerFS' incorrectly implements interface 'FileSystem'.
```

Unfortunately, TypeScript does not tell me what property or method is incorrectly implemented/specified, if any. Is this a bug in the compiler?

The relevant code can be checked out of the `webworker` branch of my [BrowserFS repository](https://github.com/jvilk/BrowserFS/tree/webworker).

Compiler command:

```
tsc --module amd src/backend/workerfs.ts
```

If that's too much code to work through, I can see if I can work it into a smaller test case...

Thanks in advance!

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #1734 and #933 by recomputing the type relation when we are about to issue an unelaborated error. This leads to slightly more verbose errors in some cases, but prevents us from issuing completely useless errors as seen in #1734.

My first try at implementing this had a separate cache parallel to the relational result cache in which to store the error, but the problem with that approach is that will sometimes reuse the same comparison (e.g. is a class assignable to the interface it implements) with different error messages. Since the diagnostic info stores the formatted message rather than the constituent nodes, we would sometimes end up issuing the wrong error message (e.g. 'foo is not assignable to bar' rather than 'foo does not implement bar').

### CODE REVIEW

**Comment by mhegazy** [comment]
:+1:

File: tests/baselines/reference/promisePermutations.errors.txt [hunk 1]
```diff
@@ -254,26 +316,41 @@ tests/cases/compiler/promisePermutations.ts(160,21): error TS2345: Argument of t
     var s10c = s10.then(testFunction10P, testFunction10, testFunction10); // ok
     var s10d = s10.then(sPromise, sPromise, sPromise); // ok
     var s10e = s10.then(nIPromise, nPromise, nIPromise); // ok
+                        ~~~~~~~~~
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/promisePermutations.errors.txt]: :-1: this is wrong

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3697,12 +3707,14 @@ module ts {
                     var maybeCache = maybeStack[depth];
                     // If result is definitely true, copy assumptions to global cache, else copy to next level up
                     var destinationCache = result === Ternary.True || depth === 0 ? relation : maybeStack[depth - 1];
-                    copyMap(/*source*/maybeCache, /*target*/destinationCache);
+                    for (var key in maybeCache) {
+                        destinationCache[key] = maybeCache[key] ? RelationComparisonResult.Succeeded : RelationComparisonResult.Failed;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Incorrect code here -- `destinationCache` is either `Map<boolean>` or `Map<RelationComparisonResult>`; note that a write of a `RelationComparisonResult` is an unsound write (and incorrect in this case)

**Comment by JsonFreeman** [comment]
:+1:

**Pull request merged**